### ISSUE TITLE
Feature request: `%tags%` Variable

### ISSUE THREAD

**Issue by infinitelyintrigued** [RealRaven2000/SmartTemplate4#320]
I use Thunderbird tags to categorize emails. I would like to see a variable (e.g., %tags%) to embed assigned tag names in templates (e.g., tag1,tag2,....)

It would also be great if the variable can do pattern matching optionally (e.g, %tags("pattern")% for matching and  %tags("^pattern") non-matching) to include and exclude some tags.

I use the same tag names in other applications that have their own email addresses. Being able to tag reply and forward messages will be a great time-saver.

Thank you

**Comment by RealRaven2000** [comment]
Maybe something like this

```
%tags()%                 -- (return all tags - comma separated)
%tags(matchPattern)%        -- returns matching tags (regular expression)
%tags(matchPattern,excludePattern)%   -- returns matching tags, exluding matches of 2nd expression
%tags(matchPattern,excludePattern,toclipboard)%  --literal parameter to copy the list to the clipboard.
```

I think it could also be useful to optionally restrict to just one tag somehow, if there was some processing or a field that couldn't take multiple tags?

**Comment by infinitelyintrigued** [comment]
Looks good to me. I can't wait to see this new feature!

**Comment by RealRaven2000** [comment]
Here is a first test version. Currently only supports 2 uses `%tags%` and `%tags(color)%` [this one only makes sense in HTML emails, just a little thing I added for better visibblity]  - try it out!

[smartTemplate-fx-4.8.2pre24.zip](https://github.com/user-attachments/files/17002943/smartTemplate-fx-4.8.2pre24.zip)

----

To install version above download zip file and drag the file into Thunderbird Add-ons Manager (do not extract contents, it won't install like that)

**Comment by infinitelyintrigued** [comment]
Yes, it works. Very nice. This will help me a lot!

I have one more request if you don't mind. Could you also provide an option to format each tag name? Some applications I use automatically recognize a text as a tag when it's prefixed with a special character such as "#". For example, %tags produces "tag1,tag2,tag3,..." and %tags(format:"#\0")% produces "#tag1,#tag2,#tag3"?

**Comment by RealRaven2000** [comment]
> I have one more request if you don't mind. Could you also provide an option to format each tag name? Some applications I use automatically recognize a text as a tag when it's prefixed with a special character such as "#". For example, %tags produces "tag1,tag2,tag3,..." and %tags(format:"#\0")% produces "#tag1,#tag2,#tag3"?

I will look into it. At the moment I am just wrapping the whole lot like this:

```js
      for (let t of tags) {
        const labelObject = `<span style="color:${t.color}; border: 1px solid ${t.color};border-radius: 0.2em; padding: 0.05em 0.1em;">${t.label}</span>`;
        labelAtoms.push(labelObject);
      }
```

I already thought to myself that the formatting string (which is now hard coded) could be somehow configurable in a global preferences somewhere.

**Comment by infinitelyintrigued** [comment]
Thank you. I'll be looking forward to it. The current implementation is very helpful as it is. I can feel my workflow is smoother.

**Comment by RealRaven2000** [comment]
I think this might be a good candidate to tie to one of the license levels either SmartTemplates Pro only or at  least ST Standard with some restrictions. Simply making it a Pro feature would be my preference, as it is a little outside the wheel house of the ordinary user. I have to make sure that I convert a few more users to licensees as I am putting a lot of work in which isn't currently reflected in sales. Partly a problem of it being too complex of course, but that's to be addressed in a different topic.

**Comment by infinitelyintrigued** [comment]
I think it makes sense to make it a pro-only feature.

**Comment by RealRaven2000** [comment]
Here is a new version with more features:

[smartTemplate-fx-4.8.2pre36.zip](https://github.com/user-attachments/files/17035183/smartTemplate-fx-4.8.2pre36.zip)

you can add the following parameters:

`color`  this will use a preset format with colored labels (as before)
`delimiter="-"`  what to put in between the tags. If this is not given as parameters it defaults to `, ` (a single comma followed by a single space)
`format="formatString"`  how to format each tag label
`toclipboard` copies the result to the clipboard instead of inserting it

_formatString_ can contain single or  multiple instances of `$label$` and `$color$`

`$label$` placeholder for the tag text
`$color$`  6 digit hex value of the tag's color

Example:
```
%tags(color,delimiter=" ")%
```

result (used in a standard quote header with small font):
![image](https://github.com/user-attachments/assets/e168e9e7-afc4-4b35-b84d-362dce44258a)

---

To install the above version download zip file and drag the file into Thunderbird Add-ons Manager (do not extract contents, it won't install like that)

**Comment by RealRaven2000** [comment]
Had to edit the previous comment, because MarkDown messed with my `$label$` and `$color$` parameters (by removing the `$` signs and turning them into italic text instead).

**Comment by RealRaven2000** [comment]
> I have one more request if you don't mind. Could you also provide an option to format each tag name? Some applications I use automatically recognize a text as a tag when it's prefixed with a special character such as "#". For example, %tags produces "tag1,tag2,tag3,..." and %tags(format:"#\0")% produces "#tag1,#tag2,#tag3"?

with the latest version I linked above, please try:

```
  %tags(format="#$label$",delimiter=",")%
```

**Comment by infinitelyintrigued** [comment]
I have this code in my template. (I copied/pasted your sample code.)
```
<b>Tags</b>: %tags(format="#$label$",delimiter=",")%
```
The tags are rendered like this:
```
#tag1#tag2#tag3
```
without the comma delimiter.

I also could not get the color parameter to work right. Could you show me a sample code with the format and delimiter parameters?

But I'm already very excited about this release!

**Comment by RealRaven2000** [comment]
> The tags are rendered like this:
> 
> ```
> #tag1#tag2#tag3
> ```
> 
> without the comma delimiter.

Looks like there is a problem with the comma (it is swallowed by the parser that breaks up the parameters, which are also delimited by commas. Had to write a fix and you will also have to escape the comma with a \ backslash. Like below; I also tried the second way with some color to stress test it, and had to make it more robust because of the "=" within the second format string:

```
%tags(format="#$label$",delimiter="\,")%
or
%tags(format="<span style='background-color:$color$;color:white;border-radius:0.2em;'>#$label$</span>",delimiter="\,")%
```

the second one looks like this:
![image](https://github.com/user-attachments/assets/89cbf8f2-52e5-4441-a501-91cbbd82ceca)

New version. [Make sure to put `\,` instead of `,` for the delimiter]:

[smartTemplate-fx-4.8.2pre41.zip](https://github.com/user-attachments/files/17037002/smartTemplate-fx-4.8.2pre41.zip)

---

To install the above version download zip file and drag the file into Thunderbird Add-ons Manager (do not extract contents, it won't install like that)

**Comment by infinitelyintrigued** [comment]
I got the first example (the comma delimiter) to work correctly. This makes a big difference.


The second color example turned out to be tricky. I use Visual Studio code to edit templates. The Prettier formatter is turned on by default. This formatting engine breaks up a long line into multiple lines. The second color example is a long line. The Prettier broke it up into multiple lines when the template was saved. This also broke %tags%. I edited the template in another text editor and restored it into a single line. Then %tags% worked.

Another thing I noticed was that my Thunderbird (128.2.1esr (64-bit) on Pop!_OS 22.04) could not render the colors. Consequently, the result was the same as the first example's. However, Firefox was able to render correctly after I saved the rendered message as an HTML file and opened it in Firefox.

**Comment by RealRaven2000** [comment]
> Another thing I noticed was that my Thunderbird (128.2.1esr (64-bit) on Pop!_OS 22.04) could not render the colors. Consequently, the result was the same as the first example's. However, Firefox was able to render correctly after I saved the rendered message as an HTML file and opened it in Firefox.

I am wondering if you have plain text only emails enabled? The HTML (with inline CSS) should display everything as expected as long as you are editing in HTML mode. Although even if I have the command in my account template and force replying in plain text (holding down Shift while clicking Reply) I still see the tags in color`, albeit with monospaced font:

![image](https://github.com/user-attachments/assets/911c1f8b-04ea-4f3a-b23c-859b53d25a5f)

so it's a little strange, can you look at the generated markup in Composer (I use ThunderHTMLEdit)?
you can also investigate the HTML in Composer by highlighting portions of the text (best more than a line or across tag boundaries) and selecting insert >> html from the menu - this will display the HTML of the selected text:

![image](https://github.com/user-attachments/assets/c73d0e1c-6a19-4fbe-9a93-985caa39d7e9)

In the meantime I also added support for the `[[optional section]]` syntax for tags, with this Composer can leave out a line if there are no tags in the replied / follow-up email. For instance:

`[[<p> Tags: %tags(color)% </p>]]`

This will make a paragraph with tags but only if at least on exists. otherwise it will not insert the paragraph. I have similar functionality for all address headers, such as cc, bcc, which can generate nicer, more concise quote headers.

[smartTemplate-fx-4.8.2pre44.zip](https://github.com/user-attachments/files/17059936/smartTemplate-fx-4.8.2pre44.zip)

----

To install the above version download zip file and drag the file into Thunderbird Add-ons Manager (do not extract contents, it won't install like that)

**Comment by RealRaven2000** [comment]
And I also added an optional text transformation parameter (use one of `uppercase`,`lowercase`,`capitalize`,`camelcase`) for the tag labels.

[smartTemplate-fx-4.8.2pre46.zip](https://github.com/user-attachments/files/17060736/smartTemplate-fx-4.8.2pre46.zip)

Example:

```html
[[<p> Tags: %tags(format="<span style='font-size:0.8em;background-color:$color$;color:white;border-radius:0.2em;padding-inline:0.1em;'>$label$</span>",delimiter=" ",uppercase)% </p>]]
```

----
To install the above version download zip file and drag the file into Thunderbird Add-ons Manager (do not extract contents, it won't install like that)

**Comment by infinitelyintrigued** [comment]
The color issue may be the way my Thunderbird is configured. It is configured to use HTML, but it seems to ignore any color-related tags. The tags are properly colored when I viewed them in other mail clients.

The optional section and the text transformation parameters are very good ideas. The text transformation can keep the tag spelling more consistent.

I started to create more tags to take advantage of this new SmartTemplates feature. It's becoming much more than I originally envisioned. I thank you. (By the way, do you know how I can remove tags from a search result? I found messages with old tags using Ctrl+Shift+F, but there is no option to remove tags from the found messages.  I want to remove old tags I don't use anymore, but I cannot find a way to do this in Thunderbird or an addon that can do this. I know it's unrelated to SmartTemplates, but you're a long-time Thunderbird user/developer. I just wondered if you do.)

**Comment by RealRaven2000** [comment]
I did some more refinements and added additional formatting switches. Latest version:

[smartTemplate-fx-4.8.2pre61.zip](https://github.com/user-attachments/files/17150180/smartTemplate-fx-4.8.2pre61.zip)

**Optional parameters**

`color`  - a preset with a colorful label (using the color of the label for font and border)
`color-filled` - another preset with a colored label (using the color of the label for background, with white font)
`filled` - tags with white font, filled with dark color 
`dark-filled` - tags with black font, filled with white color - useful for dark backgrounds.
`size=fontSizeValue`  - font size, _fontSizeValue_ is either a simple decimal numeric such as 0.8, 1.0, 1.2 etc. and will be multiplied with the standard size of em, or a full value of number+unit, eg 100px, 0.6em, 14pt etc.
`delimiter="delimitString"` - _delimitString_ = what to put in between the tags. commas need to be prefixed with a backslash `\`. If this is not given as parameters it defaults to ,  (a single comma followed by a single space)
`format="formatString`" how to format each tag label
`toclipboard` copies the result to the clipboard instead of inserting it
`uppercase` - converts label texts to uppercase
`lowercase` - convert label texts to lowercase
`capitalize `- capitalizes all label texts to (first letter each word in uppercase, the rest in lowercase)
`camelcase` - only converts the first letter to lowercase, leaves any internal mixed case intact

_formatString_ is designed as either plain text or html snippet and can contain single or multiple instances of the placeholders `$label$` and `$color$`. Make sure to enclose it in double quotes and escape any commas within `\,`:

`$label$` placeholder for the tag text
`$color$` 6 digit hex value of the mail tag's color

I chose to decrease the font sizes for the "graphical" preset values to better fit within the normal text flow. This can be overridden by adding the `size` parameter. The only question here is whether we should use bold font style and whether to add this as an optional parameter (just like size). It can spiral out of control with the amount of different possible parameters. The other option would be an additional `style` parameter to inject additional rules (such as `style=font-weight:bold`)

Here is a template block with examples:

```html
[[<p>tags: %tags%</p>]]
[[<p>tags(color,size): %tags(color,size=1.0 )%</p>]]
[[<p>tags(color-filled,uppercase,delimiter=" "): %tags(color-filled,uppercase,delimiter=" ")%</p>]]
[[<p>tags(filled,lowercase,delimiter=" - "): %tags(filled,lowercase,delimiter=" - ")%</p>]]
[[<div style="width:50%; background-color:#333; color:white; padding:0.5em; margin-block:0.2em;">tags(dark-filled): %tags(dark-filled)%<br></div>]]
[[<p>tags(format): %tags(format="<span style='background-color:$color$;color:white;border-radius:0.7em;padding-inline:0.5em'>#$label$</span>")%</p>]]
```

...and the resulting output in Composer:

![image](https://github.com/user-attachments/assets/715b3d56-a4af-4ab9-911a-959b11493d4f)

**Comment by RealRaven2000** [comment]
Latest version. I added documentation in the Variables tab (fully localised); includes a section with examples:
![image](https://github.com/user-attachments/assets/0e6f88e6-2ec9-403d-a156-825d67edcfd8)

I think I may use the Examples table as a  pattern in other sections of the variables tab, it might be useful?

[smartTemplate-fx-4.8.2pre64.zip](https://github.com/user-attachments/files/17167758/smartTemplate-fx-4.8.2pre64.zip)

---

To install the above version download zip file and drag the file into Thunderbird Add-ons Manager (do not extract contents, it won't install like that)

**Comment by infinitelyintrigued** [comment]
Examples are always good for the users to get the most out of the feature. The more powerful the feature, the more examples are needed.

**Comment by RealRaven2000** [comment]
> Examples are always good for the users to get the most out of the feature. The more powerful the feature, the more examples are needed.

Agreed, just having them fully localized into 20 languages is a huge amount of work. I think here it's kind of ok because the examples work for any language (the tags are always the same) - just the principle is shown.

![image](https://github.com/user-attachments/assets/639492df-3f35-4195-a4bf-6d32c5ee4587)

note: I need different table colors in dark mode!

**Comment by infinitelyintrigued** [comment]
This feature is working very well for me. Thank you again for your work!

**Comment by RealRaven2000** [comment]
Implemented in 4.9 - published 11/11/2024
